src/peft/utils/merge_utils.py

```diff
@@ -0,0 +1,237 @@
+
+    Args:
+        tensor (`torch.Tensor`):The tensor to get the mask from.
+        method (`str`):The method to use to get the mask. Should be one of ["total", "frequency"].
+
+    Returns:
+        `torch.Tensor`: The majority sign mask.
+    """
+
+    sign = tensor.sign()
+    if method == "total":
+        sign_magnitude = (sign * tensor.abs()).sum(dim=0)
+    elif method == "frequency":
+        sign_magnitude = sign.sum(dim=0)
+    else:
+        raise RuntimeError(f'Unimplemented mask method "{method}"')
+    majority_sign = torch.where(sign_magnitude >= 0, 1, -1)
+    return sign == majority_sign
+
+
+def disjoint_merge(task_tensors: torch.Tensor, majority_sign_mask: torch.Tensor) -> torch.Tensor:
+    """
+    Merge the task tensors using disjoint merge.
+
+    Args:
+        task_tensors (`torch.Tensor`):The task tensors to merge.
+        majority_sign_mask (`torch.Tensor`):The mask of the majority sign across the task tensors.
+
+    Returns:
+        `torch.Tensor`: The merged tensor.
+    """
+    mixed_task_tensors = (task_tensors * majority_sign_mask).sum(dim=0)
+    num_params_preserved = majority_sign_mask.sum(dim=0)
+    return mixed_task_tensors / torch.clamp(num_params_preserved, min=1.0)
+
+
+def task_arthimetic(task_tensors: torch.Tensor, weights: torch.Tensor) -> torch.Tensor:
+    """
+    Merge the task tensors using `task arthimetic`.
+
+    Args:
+        task_tensors(`torch.Tensor`):The task tensors to merge.
+        weights (`torch.Tensor`):The weights of the task tensors.
+
+    Returns:
+        `torch.Tensor`: The merged tensor.
+    """
+    task_tensors = torch.stack(task_tensors, dim=0)
+    # weighted task tensors
+    weights = reshape_weight_task_tensors(task_tensors, weights)
+    weighted_task_tensors = task_tensors * weights
+    mixed_task_tensors = weighted_task_tensors.sum(dim=0)
+    return mixed_task_tensors
+
+
+def ties(
+    task_tensors: torch.Tensor,
+    weights: torch.Tensor,
+    density: float,
+    majority_sign_method: Literal["total", "frequency"] = "total",
+) -> torch.Tensor:
+    """
+    Merge the task tensors using `ties`.
+
+    Args:
+        task_tensors(`torch.Tensor`):The task tensors to merge.
+        weights (`torch.Tensor`):The weights of the task tensors.
+        density (`float`):The fraction of values to preserve. Should be in [0,1].
+        majority_sign_method (`str`):
+            The method to use to get the majority sign mask. Should be one of ["total", "frequency"].
+
+    Returns:
+        `torch.Tensor`: The merged tensor.
+    """
+    # sparsify
+    task_tensors = [prune(tensor, density, method="magnitude") for tensor in task_tensors]
+    task_tensors = torch.stack(task_tensors, dim=0)
+    # weighted task tensors
+    weights = reshape_weight_task_tensors(task_tensors, weights)
+    weighted_task_tensors = task_tensors * weights
+    # Elect Sign
+    majority_sign_mask = calculate_majority_sign_mask(weighted_task_tensors, method=majority_sign_method)
```

Comment:
Important: I am not why is the majority sign computed on the `weighted_task_tensor` this can have an unintended sign selection because the user can pass very high weights for some model. The pruning and majority sign selection steps are independent of the weights that are used to merge the model. In the Ties paper, we pruned, then selected majority sign, did disjoint merge to obtain the merged task vector $\tau_m$ and then used a single weight to $\lambda$ to scale $\tau_m$ and add back to the pretrained model. 

I understand that you might want to provide an option to use different weights for different task vectors when merging but still the sign selection should happen before the task vectors are being rescaled. 

Am I missing something?

<img width="418" alt="image" src="https://github.com/huggingface/peft/assets/15224633/f0d0effa-f16d-49c3-82a6-1ef1d24c3786">

Reply:
Done, changed the ordering such that weighing happens after getting the majority sign mask.

Reply:
I am not sure if I am looking at something wrong but the changes are not reflected for me.  Also the should also be done in Dare_ties.

![image](https://github.com/huggingface/peft/assets/15224633/7b7139e5-2539-423f-beb6-2ce5067e7af8)

The order should be, `prune -> majority sign mask -> weighing -> disjoint merge`